test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -444,6 +484,495 @@ public void testDurableDataLogFail() throws Exception {
+                .controllerURI(controllerStarter.controllerURI).build());
+        clientFactory = new ClientFactoryImpl(SCOPE, controllerStarter.controller, connectionFactory);
+        readerGroupManager = new ReaderGroupManagerImpl(SCOPE, controllerStarter.controller, clientFactory);
+
+        // Try creating the same segments again with the new controller
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM1);
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM2);
+
+        // Get reader group.
+        readerGroup1 = readerGroupManager.getReaderGroup(testReaderGroup1);
+        readerGroup2 = readerGroupManager.getReaderGroup(testReaderGroup2);
+        Assert.assertNotNull(readerGroup1);
+        Assert.assertNotNull(readerGroup2);
+
+        reader1 = clientFactory.createReader(testReader1, testReaderGroup1, new UTF8StringSerializer(), ReaderConfig.builder().build());
+        reader2 = clientFactory.createReader(testReader2, testReaderGroup2, new UTF8StringSerializer(), ReaderConfig.builder().build());
+
+        // Read rest of the events.
+        readNEvents(reader1, TOTAL_NUM_EVENTS - NUM_EVENTS);
+        readNEvents(reader2, TOTAL_NUM_EVENTS - NUM_EVENTS);
+
+        // Reading next event should return null.
+        assertNull(reader1.readNextEvent(5000).getEvent());
+        assertNull(reader2.readNextEvent(5000).getEvent());
+        reader1.close();
+        reader2.close();
+    }
+
+    /**
+     * Tests the data recovery scenario with watermarking events.
+     *  What test does, step by step:
+     *  1. Starts Pravega locally with just 4 segment containers.
+     *  2. Writes 300 events to two different segments with watermarks.
+     *  3. Waits for all segments created to be flushed to the long term storage.
+     *  4. Shuts down the controller, segment store and bookeeper/zookeeper.
+     *  5. Deletes container metadata segment and its attribute segment from the old LTS.
+     *  6. Starts 4 debug segment containers using a new bookeeper/zookeeper and the old LTS.
+     *  7. Re-creates the container metadata segment in Tier1 and let's it flushed to the LTS.
+     *  8. Starts segment store and controller.
+     *  9. Read all events and verify that all events are below the bounds.
+     * @throws Exception    In case of an exception occurred while execution.
+     */
+    @Test(timeout = 180000)
+    public void testDurableDataLogFailRecoveryWatermarking() throws Exception {
+        int instanceId = 0;
+        int containerCount = 4;
+        String scope = "scopeTx";
+        String stream = "streamTx";
+
+        // Creating a long term storage only once here.
+        this.storageFactory = new InMemoryStorageFactory(executorService());
+        log.info("Created a long term storage.");
+
+        // Start a new BK & ZK, segment store and controller
+        this.bookKeeperStarter = setUpNewBK(instanceId++);
```

Comment:
Same comments as in the previous tests.

Reply:
included a `setUpPravega` method which instantiates a bookeeper, segmentstore and controller with given bookie count, container count and instanceId.